### PR TITLE
Add Photon Geocoder

### PR DESCRIPTION
**Pull request by amenk**
Adds the photon geocoder

See https://github.com/opentripplanner/otp-react-redux/issues/706

Working as a proof of concept within otp-react-redux for autocomplete + reverse geocoding

[ ] Add search type (currently only autocomplete and reverse were tested)
[ ] Add bbox support
[ ] Support base URL overwrite (?)
[ ] Add more tests (?)

Also fixes a typo in the console.err message.

### CODE REVIEW

**Comment by amenk** [comment]
@miles-grant-ibigroup nice you are reviewing it already - if you want you can tidy/finalize it, you can --- otherwise I will try to work as soon as possible on your incoming comments, probably at the next weekend :-)

File: packages/geocoder/src/apis/photon/index.ts [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+ * @return {Promise}                            A Promise that'll get resolved with search result
+ */
+function reverse({ options, point }: ReverseQuery): Promise<PhotonResponse> {
+  const query: PhotonQuery = {
+  };
+
+  if (point) {
+    const { lat, lon }: LonLatOutput = normalize(point);
+    query.lat = `${lat}`;
+    query.lon = `${lon}`;
```
Comment (miles-grant-ibigroup) [inline packages/geocoder/src/apis/photon/index.ts:149]: This can probably be simplified a bit

**Review by miles-grant-ibigroup — APPROVED** [review]
Have tested this extensively and it's a great addition to otp-ui. Thanks for letting us know about the geocoder and thanks for implementing! 

A few small nitpicks -- if the second reviewer doesn't have any major changes I should have time to make the changes myself before the weekend!

**Comment by amenk** [comment]
I am not sure if I implemented also the "search" -- I tested only reverse geocoding and autocomplete via otp-react-redux

**Comment by miles-grant-ibigroup** [comment]
in otp-rr search is used when you hit enter while typing something up! It worked in my testing

**Comment by amenk** [comment]
Thanks again for the positive feedback.

Also was not sure if the configuration of an alternative base URL works - the default is Komoot's public instance, but one might also want to use a self-hosted.

**Review by binh-dam-ibigroup — COMMENTED** [review]
Looks good overall, but should the reverse query miss the longitude in the response?

File: packages/geocoder/src/geocoders/photon.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    if (this.geocoderConfig?.reverseUseFeatureCollection) {
+      response.features.forEach((value) => {
+        value.properties.label = generateLabel(value.properties);
+      })
+      return response
+    }
+
+    const { lat, lon } = response.point;
+
+    const firstFeature = response.features[0];
```
[the remark this answers is not in the document]
Comment (amenk) [reply packages/geocoder/src/geocoders/photon.ts:84]: Good catch, fixing this ...

File: packages/geocoder/src/geocoders/photon.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+  const propertyList = [];
+  ["name", "street", "district", "state", "postcode", "city", "country"].forEach((propertyName) => {
+        if (typeof properties[propertyName] === "undefined") {
+          return;
+        }
+        const value = `${properties[propertyName]}`;
+        if (value.length > 0) {
+          propertyList.push(value);
+        }
+      }
```
[the remark this answers is not in the document]
Comment (amenk) [reply packages/geocoder/src/geocoders/photon.ts]: fixed

File: packages/geocoder/src/geocoders/photon.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    return {
+      boundary,
+      focusPoint,
+      options,
+      // explicitly send over null for sources if provided sources is not truthy
+      // in order to avoid default isomorphic-mapzen-search sources form being
+      // applied
+      sources: sources || null,
+      url: baseUrl ? `${baseUrl}/search` : undefined,
+      format: false, // keep as returned GeoJSON,
```
[the remark this answers is not in the document]
Comment (amenk) [reply packages/geocoder/src/geocoders/photon.ts]: I copied this from Here, so it should be consistent. Shall we sort them all?

[the remark this answers is not in the document]
Comment (binh-dam-ibigroup) [reply packages/geocoder/src/geocoders/photon.ts]: Not a deal breaker, but it is a nice to have for new and modified code.

[the remark this answers is not in the document]
Comment (amenk) [reply packages/geocoder/src/geocoders/photon.ts]: should be resolved already, as I removed some not needed properties here it became sorted magically :)

File: packages/geocoder/src/index.story.tsx [hunk 5]
```diff
@@ -48,6 +49,14 @@ const GeocoderTester = ({
     reverseUseFeatureCollection
   });
 
+  const photonGeocoder = getGeocoder({
+    baseUrl: "https://photon.komoot.io",
+    focusPoint,
+    size: 1,
+    type: "PHOTON",
+    reverseUseFeatureCollection
```
[the remark this answers is not in the document]
Comment (amenk) [reply packages/geocoder/src/index.story.tsx:57]: sorted.

**Comment by amenk** [comment]
If this is merged, will otp-react-redux also be updated?

**Comment by miles-grant-ibigroup** [comment]
Thanks for all your changes! Sadly otp-rr is not updated when this is merged, but don't worry about that we can take care of making the package update PR!

**Comment by amenk** [comment]
Okay, are the tests flaky or did I break something?

**Comment by miles-grant-ibigroup** [comment]
First time I've seen this issue but it doesn't seem to be your fault...

**Comment by amenk** [comment]
Now the checks are green :-) @binh-dam-ibigroup can you review again?

**Review by binh-dam-ibigroup — APPROVED** [review]
Thanks for the work!

**Comment by amenk** [comment]
Awesome, thanks. Do you make the otp-ui upgrade of the module or shall I?

**Comment by miles-grant-ibigroup** [comment]
It's done automatically by a bot!

**Pull request merged**